#159762182 Integrate Payment

yomigeek opened the pull request:
#### What does this PR do?
Integrate Payement System to allow users upgrade to a premium account on the platform

#### Description of Task to be completed?
- consume Flutterwave Rave API
- create a dummy front-end page to initiate the payment
- create a middleware that checks if the payment reference is provided 
- prevent user account upgrade if payment reference is invalid during a check with the Flutterwave Rave API
- upgrades user account if payment reference is valid during a check with the Flutterwave Rave API

#### How should this be manually tested?
 - clone the repo: Run `git clone https://github.com/andela/elven-ah.git` in the folder where you want the repo to be saved.
 - navigate into the newly created folder and install the dependencies: `cd elven-ah && npm install`
 - copy .env.example to .env: `cp .env.example .env`
 - provision a PostgreSQL database and add the credentials to the .env file.
 - start the server using `npm start`
- initiate the payment using a front-end client to get the payment reference via the query parameter
 - upgrade the user account by making a `GET` request to `/api/pay` with `ref as a query parameter`

#### Any background context you want to provide?
- for testing purpose use the rave test API keys by sigining up with flutterwave rave sandbox
- on the front-end use the test javascript reference (https://ravesandboxapi.flutterwave.com/flwv3-pug/getpaidx/api/flwpbf-inline.js)

#### What are the relevant pivotal tracker stories?
[#159762182](https://www.pivotaltracker.com/story/show/159762182)

#### Screenshots (if appropriate)\
N/A


#### Questions:
N/A

[Finishes #159762182]

yomigeek committed:
feat(integrate-payment): integrate payment on the platform
- add a sketchy UI for checking of the process
- add a payment model
- add a payment controller
- install the flutterwave rave package

yomigeek committed:
feat(integrate-payment): integrate payment on the platform
- add a sketchy UI for checking of the process
- add a payment model
- add a payment controller
- install the flutterwave rave package

yomigeek committed:
feat(integrate-payment): integrate payment on the platform

- modify payment controller file
- modify payment model

yomigeek committed:
feat(integrate-payment): integrate payment on the platform

- modify payment controller file
- modify payment model

madeofhuman committed:
fix(comment): fix app crash on comment reply with invalid parentId

- create comment reply as first level comment when invalid parentId is supplied

[Delivers #160216041]

madeofhuman committed:
chore(test): cover new comment reply feature edge case

- add test for comment reply feature that creates comment reply with invalid parent id as first level comment
- refactor comment tests to remove unnecessary fields in request body

[Finishes #160216041]

madeofhuman committed:
chore(refactor): refactor CommentController

- refactor CommentController to use aync/await
- extract repeated slug validation code to middleware
- modify test to read updated responses

[Finishes #160216041]

madeofhuman committed:
chore(test): fix falling test coverage

- add more tests to cover more edge cases

[Finishes #160216041]

yomigeek committed:
feat(integrate-payment): implement payment integration
- add more unit tests
- create a check payment refeerence middleware
- modify payment controller file

[Delivers #159762182]

yomigeek committed:
feat(integrate-payment): implement payment integration
- add more unit tests
- create a check payment refeerence middleware
- modify payment controller file

[Delivers #159762182]

yomigeek committed:
feat(integrate-payment): implement changes based on reviews
- fix hound issues
- modify test

[Delivers #159762182]

yomigeek committed:
feat(integrate-payment): implement changes based on reviews
- fix hound issues
- modify test

[Delivers #159762182]

yomigeek committed:
feat(integrate-payment): improve test coverage
- improve test coverage removing redundant codes

[Delivers #159762182]

yomigeek committed:
feat(integrate-payment): improve test coverage
- improve test coverage removing redundant codes

[Delivers #159762182]

Veraclins commented on server/controllers/PaymentController.js:
```diff
@@ -0,0 +1,79 @@
+      .then((response) => {
+        const { id: userId, email } = req.user;
+        const transactionValues = response.body.data;
+        const subscriptionType = transactionValues.amount === SUBSCRIPTION_RATE ? 'month' : 'annual';
```
@yomigeek nice job here and quite expressive. I feel the SUBSCRIPTION_RATE constant should be MONTHLY_SUBSCRIPTION_RATE to add context and make the comparable unambiguous.

yomigeek replied:
Thanks for this notification, I would implement this change..

madeofhuman committed:
Merge pull request #29 from andela/fix/160216041/handle-invalid-commentId-error

#160216041 fix app crash on comment reply creation with invalid parentId

Veraclins commented on server/controllers/PaymentController.js:
```diff
@@ -0,0 +1,79 @@
+      });
+  }
+
+  static ProcessUpgrade(paymentInfo, res) {
```
Is there a reason why there is a ProcessUpgrade method and no method to create a payment? I see that you are actually creating the payment if none is found and updating otherwise. Don't you think ProcessUpgrade may not really be the best method name? I believe processPayment could be a better name. That could include both creating and upgrading the payment and still be descriptive of the whole action. Also, your method name is not in camelCase.

yomigeek replied:
My bad as regards the camelCasing issue, I will implement the change. As regards the method name, I used that because the whole process is all about upgrading a user from a free account to a premium account.

Oluwafayokemi reviewed (approved):
Nice job! @yomigeek. Well done

Veraclins commented on server/middlewares/checkReference.js:
```diff
@@ -0,0 +1,20 @@
+  * @param {object} next the next middleware function
+
+  */
+const checkReference = (req, res, next) => {
```
Is there a reason why this is implemented as a middleware instead of as a method in the controller? Given that it is used in just a single place I feel that it is better as a method in the PaymentController. This way everything about the payment is contained in a single place. (Single Responsibility Principle).

yomigeek replied:
Okay, I did that to avoid longer codes to in the validate controller, more or less like every aspect having its own purpose within the whole process.

Veraclins reviewed (changes requested):
@yomigeek nice job here. Just check out my comments and implement the feedback where necessary.

yomigeek committed:
feat(integrate-payment): Implement modification to payment integration
- add new models based on recent modification to accomodate article subscription
- modify the user and article model to accomodate subscription and updates
- modify the payment controller file

[Delivers #159762182]

yomigeek committed:
feat(integrate-payment): Implement modification to payment integration
- add new models based on recent modification to accomodate article subscription
- modify the user and article model to accomodate subscription and updates
- modify the payment controller file

[Delivers #159762182]

yomigeek committed:
feat(integrate-payment): Implement modifications to payment integration
- add more units
- modify payment controller

[Delivers #159762182]

yomigeek committed:
feat(integrate-payment): Implement modifications to payment integration
- add more units
- modify payment controller

[Delivers #159762182]

Veraclins reviewed (changes requested):
@yomigeek you have done a nice job here. I particularly love the way you arranged the methods in the order of their invocation. This makes it very easy to understand the flow of operation. There are, however, a few areas I feel you may want to look into.
1. One of the recommendations for writing clean codes is to use verbs (or actions) for method names. This you did well with the checkPaymentType method, however, I feel you can reword the other method names. For instance, accountUpgrade could be upgradeAccount, accountUpgradeProcessor could be processAccountUpgrade, etc.
2. Since all the failed payment responses return 400 status code, don't you think that defining them everywhere is rather an unnecessary repetition? I believe you can use the number (like you did for success responses) in the method itself.
3. On line 111, I see that you are checking that an articleId is supplied but you are not checking that the supplied id is a number. This could cause database errors which would break the app.
4. The userSubscriptionDetailUpdate method was declared to take two arguments but you called it on line  75 without any argument. How does that work?
5. Finally, you may want to check the comments (docs) for successPaymentResponse and failedPaymentResponse. From their description in the comments, they seem to be doing exactly what userSubscriptionDetailUpdate does.
Once again, nice job bro.

yomigeek committed:
feat(integrate-payment): implement changes based on peer reviews
- add more unit tests
- modify payment controller file

[Deliver #159762182]

yomigeek committed:
feat(integrate-payment): implement changes based on peer reviews
- add more unit tests
- modify payment controller file

[Deliver #159762182]

yomigeek commented:
@Veraclins Thanks for the reviews you gave earlier. Reviews and changes have been implemented as deemed necessary...

madeofhuman committed:
#160127137 version the API (#31)

chore(readme): add api docs link and update url

- add link to the API docs and update the url on the readme file to reflect the versioning

[Finishes #160127137]

madeofhuman reviewed (commented):
You did a really great work here, Yomi, especially given the changing scope of the task. Well done! Just address the conflicts arising from my PR merge and we're golden.

yomigeek committed:
feat(integrate-payment): implement changes to acommodate api versioning
- make changes to files and folder structures to acoomodate api versioning

[Delivers #159762182]

yomigeek committed:
feat(integrate-payment): fix merge issues
- fix merge issues on branch

[Delivers #159762182]

tonymontaro reviewed (approved):
Excellent work!
